### ISSUE TITLE
EfficientNet not modified as described in the EfficientDet paper

### ISSUE THREAD

**Issue by rmcavoy** [signatrix/efficientdet#29]
Unless I am misreading your code, this EfficientDet implementation is using the unmodified EfficientNet implementation from LukeMelas's code. The base EfficientNet as described in the original EfficientNet paper is actually slightly different than the one used in the EfficientDet paper (though the authors don't do a good job of conveying this clearly). The original EfficientNet  has a stride of 1 for the 5th and 7th feature level (table 1 of the efficientNet paper) resulting in the input feature size needing to be divisible by 32 (5 feature size decrease). The EfficientNet from the EfficientDet paper has a stride of 2 for every feature level (Figure 3 of the EfficientDet paper) which is why they say the resolution needs to a multiple of 128  (7 feature size decreases) and why the diagram shows P/2^n for the feature scale.

I am guessing you may be aware of this due to the non-standard convolutions "conv3-7" correcting for the fact that you only extract 3 features from the backbone instead of the 5 described in the original paper but it would be good if you mentioned this in code comments and/or the README.

**Comment by florisdesmedt** [comment]
Hmm, I missed that. Any idea how this could influence accuracy? The current implementation does not use features from the end of the network, so I would expect some influence.

I'm trying a quick hack by changing the stride in block_8 and block_15 and adapting the conv3-7 to all do a 1x1 convolution. Maybe it would be better to retrain the backbone to cope with the new stride? 

Since I'm testing on a single GPU at the moment, it could take some time before I can see the impact of this change.

**Comment by Shi-Xiaoyu** [comment]
Hi, I am struggling with this point two. I see that most implementations are using the same method as this one(i.e., add two more stride=2 convs before bifpn to get 1/64 and 1/128 features). And it works!? But I feel this process loses many information since the input channel of the second conv is only 64 for D0.
But the author of the paper doesn't explicitly point out that whether they change the backbone or not.

**Comment by florisdesmedt** [comment]
I agree some information is lost by reducing to 64 channels. But this is part of the detector design, independent of changing the backbone or not. Also take into account that reducing the layers has the biggest impact on the last feature layers, which are relatively very large for object detection. e.g 1/128 results in a 4x4 feature map on input size 512x512 of D0.
Currently I'm considering leaving the backbone as is, extracting more feature layers (although at incorrect scale) and fix the scaling with stride operations per feature layer similar to the current implementation.

**Comment by rvandeghen** [comment]
Also `self.conv5 = nn.Conv2d(192, self.num_channels, kernel_size=1, stride=1, padding=0)` should be 
self.conv5 = nn.Conv2d(112, self.num_channels, kernel_size=1, stride=1, padding=0) according to the EfficientNet baseline ?

**Comment by florisdesmedt** [comment]
The implementation takes feature layers from the EfficientNet backbone based on if they perform a scaledown (stride=[2,2]). Since the layers of performing a scaledown do not match between EfficientNet and figure3 in the paper of EfficientDet, the number of layers doesn't meet correspondingly. 192 channels should be conv6 I think.
So far I could not tell if adding extra striding steps into the backbone solves the issue.

**Comment by rmcavoy** [comment]
This can be fixed without having to stop using LukeMelas's code. You just need to make a function that passes the correct block_args in as the code was made to be flexible. This would just require making a custom from_name, model_params and load_from_pretrained. The correct block_args string from LukeMelas's code (def efficientnet in utils.py) should be changed to
    blocks_args = [
        'r1_k3_s11_e1_i32_o16_se0.25', 'r2_k3_s22_e6_i16_o24_se0.25',
        'r2_k5_s22_e6_i24_o40_se0.25', 'r3_k3_s22_e6_i40_o80_se0.25',
        'r3_k5_s22_e6_i80_o112_se0.25', 'r4_k5_s22_e6_i112_o192_se0.25',
        'r1_k3_s22_e6_i192_o320_se0.25',
    ]

**Comment by vakkov** [comment]
https://github.com/lukemelas/EfficientNet-PyTorch/issues/13

erm, nope; check the referenced link in the issue above

**Comment by rmcavoy** [comment]
@vakkov That link is referring to the original EfficientNet paper being wrong. That is an entirely separate matter from whether the EfficientDet paper used a different set of strides from the EfficientNet paper. There is as of yet no evidence they used the original EfficientNet strides in the EfficientNet and plenty of evidence that they switched out the strides to all be 2.

**Comment by vakkov** [comment]
@rmcavoy my bad, sorry. Have you tried retraining e.g. the D0 model with strides of 2 in EffNet and (i guess) fixing the conv layers of efficientDet?

**Comment by florisdesmedt** [comment]
I tried a few different configurations of fixing the strides.

- change the stride in the efficientnet model (changed the strides after reading the model)
- extracted all required layers and changed resolution using the self.convX filters (similar to https://github.com/signatrix/efficientdet/blob/master/src/model.py#L210) 

In both cases the validation accuracy stagnated around 20mAP for D0 (while the training loss still decreased). I experimented with warm-up, simulating larger batch sizes, ... but so far no luck.

**Comment by rmcavoy** [comment]
No worries. I have not tested this code in particular. I only found this error because I was searching through several repos to try and validate them versus each other. I don't even think this correction is guaranteed to change performance in away that wouldn't require very thorough testing to disambiguate. It is just a matter of whether the code implements what it says in the paper so someone testing this would know that they are working with the correct algorithm.
